Release v1.0.0-rc.4 (#6045)
ZcashFoundation/zebra: CHANGELOG.md

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org).

## [Zebra 1.0.0-rc.4](https://github.com/ZcashFoundation/zebra/releases/tag/v1.0.0-rc.4) - 2023-01-30

In this release we fixed bugs and inconsistencies between zcashd and zebrad in the output of the `getblocktemplate` RPC method. In addition, we added block proposal mode to the `getblocktemplate` RPC, while we continue the effort of adding and testing mining pool RPC methods.

### Breaking Changes

> - Strings in zebra configuration file now use double quotes, caused by upgrading the `toml` crate. Old configs will still work [#6029](https://github.com/ZcashFoundation/zebra/pull/6029)

Reviewer: I don't think it's a breaking change if old configs will still work?

Optional: put this under a "Miscellaneous" section instead.

Reviewer 2: Yeah I think I might have tagged it wrong, sorry about that!

Author: Thanks, changed at https://github.com/ZcashFoundation/zebra/pull/6045/commits/fba648ca56a5ab2e437894b78a7d109b72e091bb

- Add getpeerinfo RPC method ([#5951](https://github.com/ZcashFoundation/zebra/pull/5951))
- Add proposal capability to getblocktemplate ([#5870](https://github.com/ZcashFoundation/zebra/pull/5870))
- Add a test to check that the Docker image config works ([#5968](https://github.com/ZcashFoundation/zebra/pull/5968))
- Make `zebra-checkpoints` work for zebrad backend ([#5894](https://github.com/ZcashFoundation/zebra/pull/5894))
- Add test dependency from zebra-rpc to zebra-network with correct features ([#5992](https://github.com/ZcashFoundation/zebra/pull/5992))
- Document zebra download command ([#5901](https://github.com/ZcashFoundation/zebra/pull/5901))

### Fixed

- Return detailed errors to the RPC client when a block proposal fails ([#5993](https://github.com/ZcashFoundation/zebra/pull/5993))
- Avoid selecting duplicate transactions in block templates ([#6006](https://github.com/ZcashFoundation/zebra/pull/6006))
- Calculate getblocktemplate RPC testnet min and max times correctly ([#5925](https://github.com/ZcashFoundation/zebra/pull/5925))
- Fix Merkle root transaction order in getblocktemplate RPC method ([#5953](https://github.com/ZcashFoundation/zebra/pull/5953))

> ### Miscellaneous

Reviewer 2: The standard name for this category is "Changed":
https://keepachangelog.com/en/1.0.0/#how

Sorry I missed this before!

```suggestion
### Changed
```

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/), and this project adheres to [Semantic Versioning](https://semver.org).

## [Zebra 1.0.0-rc.4](https://github.com/ZcashFoundation/zebra/releases/tag/v1.0.0-rc.4) - 2023-01-30

> In this release we fixed bugs and inconsistencies between zcashd and zebrad in the output of the `getblocktemplate` RPC mehod. In addition, we added the proposal mode feature while we continue the effort of adding and testing mining pool RPC methods. 

Reviewer 2: Some users might not know what the proposal mode is.

```suggestion
In this release we fixed bugs and inconsistencies between zcashd and zebrad in the output of the `getblocktemplate` RPC method. In addition, we added block proposal mode to the `getblocktemplate` RPC, while we continue the effort of adding and testing mining pool RPC methods. 
```

Author: Thanks, changed in https://github.com/ZcashFoundation/zebra/pull/6045/commits/6862ce4e3dd43e5dab6f87e2d49b0069897f2d60

### Breaking Changes

> - Strings in zebra configuration file now use double quotes, caused by bumping the `toml` crate. Old configs will still work [#6029](https://github.com/ZcashFoundation/zebra/pull/6029)

Reviewer 2: Some users might not know what "bumping" means.

```suggestion
- Strings in zebra configuration file now use double quotes, caused by upgrading the `toml` crate. Old configs will still work [#6029](https://github.com/ZcashFoundation/zebra/pull/6029)
```

Author: Thanks https://github.com/ZcashFoundation/zebra/pull/6045/commits/6862ce4e3dd43e5dab6f87e2d49b0069897f2d60

### Security

> - Verify the lock times of mempool transactions ([#6027](https://github.com/ZcashFoundation/zebra/pull/6027))

Reviewer 2: Let's explain the bug, and credit the person who reported it. (Sorry I forgot this in the PR description.)

```suggestion
- Verify the lock times of mempool transactions. Previously, Zebra was ignoring mempool transaction lock times, but checking them in mined blocks. Credit to DeckerSU for reporting this issue. ([#6027](https://github.com/ZcashFoundation/zebra/pull/6027))
```

Author: https://github.com/ZcashFoundation/zebra/pull/6045/commits/6862ce4e3dd43e5dab6f87e2d49b0069897f2d60

- Avoid selecting duplicate transactions in block templates ([#6006](https://github.com/ZcashFoundation/zebra/pull/6006))
- Calculate getblocktemplate RPC testnet min and max times correctly ([#5925](https://github.com/ZcashFoundation/zebra/pull/5925))

> - Fix Merkle root transaction order in getblocktemplate RPC method ([#5953](https://github.com/ZcashFoundation/zebra/pull/5953))

Reviewer 2: Optional: I wonder if these are security issues, because they are in an experimental feature that isn't in production. What do you think?

(I don't mind either way, I just wanted to check.)

Author: I moved them to Fixed section in https://github.com/ZcashFoundation/zebra/pull/6045/commits/6862ce4e3dd43e5dab6f87e2d49b0069897f2d60

> - Bump bumpalo from 3.8.0 to 3.12.0 ([#6015](https://github.com/ZcashFoundation/zebra/pull/6015))
> - Bump libgit2-sys from 0.14.0+1.5.0 to 0.14.2+1.5.1 ([#6014](https://github.com/ZcashFoundation/zebra/pull/6014))
> - Bump tokio from 1.24.1 to 1.24.2 ([#5995](https://github.com/ZcashFoundation/zebra/pull/5995))

Reviewer 2: It might help users if we explain what the security issue is, and what its impact is.
```suggestion
- Bump bumpalo from 3.8.0 to 3.12.0, removing undefined behaviour on `wasm` targets. These targets are not supported Zebra platforms. ([#6015](https://github.com/ZcashFoundation/zebra/pull/6015))
- Bump libgit2-sys from 0.14.0+1.5.0 to 0.14.2+1.5.1, to ensure that SSH server keys are checked. Zebra only uses `libgit2` during builds, and we don't make SSH connections. ([#6014](https://github.com/ZcashFoundation/zebra/pull/6014))
- Bump tokio from 1.24.1 to 1.24.2, to fix unsoundness. The unsound methods are not directly used by Zebra. ([#5995](https://github.com/ZcashFoundation/zebra/pull/5995))
```

Author: https://github.com/ZcashFoundation/zebra/pull/6045/commits/6862ce4e3dd43e5dab6f87e2d49b0069897f2d60

### Added

- Add getpeerinfo RPC method ([#5951](https://github.com/ZcashFoundation/zebra/pull/5951))
- Add proposal capability to getblocktemplate ([#5870](https://github.com/ZcashFoundation/zebra/pull/5870))

> - Add a test to check that the Docker image config works ([#5968](https://github.com/ZcashFoundation/zebra/pull/5968))

Reviewer 2: Optional: This might be a trivial change, it doesn't impact users.

Author: Yea , thanks, there are a bunch that actually do not impact users much or at all. Removed them as suggested in https://github.com/ZcashFoundation/zebra/pull/6045/commits/6862ce4e3dd43e5dab6f87e2d49b0069897f2d60

- Make `zebra-checkpoints` work for zebrad backend ([#5894](https://github.com/ZcashFoundation/zebra/pull/5894))

> - Add test dependency from zebra-rpc to zebra-network with correct features ([#5992](https://github.com/ZcashFoundation/zebra/pull/5992))

Reviewer 2: Optional: This is a trivial change, it doesn't impact users.

Author: https://github.com/ZcashFoundation/zebra/pull/6045/commits/6862ce4e3dd43e5dab6f87e2d49b0069897f2d60

- Document zebra download command ([#5901](https://github.com/ZcashFoundation/zebra/pull/5901))

### Changed

- Update 'Design Overview' page of Zebra book ([#5892](https://github.com/ZcashFoundation/zebra/pull/5892))
- Sort transaction hashes like zcashd in getrawmempool RPC response ([#5994](https://github.com/ZcashFoundation/zebra/pull/5994))

> - Create test harness for calling getblocktemplate in proposal mode, but don't use it yet ([#5884](https://github.com/ZcashFoundation/zebra/pull/5884))

Reviewer 2: Optional: I think this doesn't impact users?

Author: https://github.com/ZcashFoundation/zebra/pull/6045/commits/6862ce4e3dd43e5dab6f87e2d49b0069897f2d60

### Fixed

- Fallback to a stable buildkit version for `docker/build-push-action` ([#5985](https://github.com/ZcashFoundation/zebra/pull/5985))
- Move docker push fix to the setup action ([#5998](https://github.com/ZcashFoundation/zebra/pull/5998))
- Remove warnings caused by missing `actions/checkout` step ([#5874](https://github.com/ZcashFoundation/zebra/pull/5874))
- Fix new lints in nightly clippy ([#5959](https://github.com/ZcashFoundation/zebra/pull/5959))
- Silence future-incompat warnings until we upgrade Abscissa ([#6024](https://github.com/ZcashFoundation/zebra/pull/6024))
- Add test dependency from zebra-rpc to zebra-network with correct features ([#5992](https://github.com/ZcashFoundation/zebra/pull/5992))

> - Hex-encode debug format of commitments, nonces, and nullifiers ([#5960](https://github.com/ZcashFoundation/zebra/pull/5960))

Reviewer 2: I don't think users can see any of these changes, so they might be trivial.

Author: https://github.com/ZcashFoundation/zebra/pull/6045/commits/6862ce4e3dd43e5dab6f87e2d49b0069897f2d60
